tests/test_limit.py

```diff
@@ -0,0 +1,139 @@
+def test_less_equal(graph, testclass):
+    def test(allow_graph=True):
+        limit = testclass(limit_fn=inside_rect_limits, rect_limits=rect_limits)
+        limit2 = testclass(limit_fn=inside_rect_limits, rect_limits=rect_limit_enlarged)
+
+        assert not limit.has_rect_limits
+        assert limit.has_limits
+        assert not limit.limits_not_set
+        assert not limit.limits_are_false
+        inside = limit.inside(2)
+        inside2 = limit.inside(2, guarantee_limits=True)
+        outside = limit.inside(4)
+
+        assert not limit2.has_rect_limits
+        assert limit2.has_limits
+        assert not limit2.limits_not_set
+        assert not limit2.limits_are_false
+        inside21 = limit2.inside(2)
+        inside22 = limit2.inside(2, guarantee_limits=True)
+        outside2 = limit2.inside(4)
+        if graph:
+            equal = limit.equal(limit2, allow_graph=allow_graph)
+        else:
+            equal = limit == limit2
+
+        return inside, inside2, outside, equal, inside21, inside22, outside2
+
+    if graph:
+        test = z.function(test)
+    inside, inside2, outside, equal, inside21, inside22, outside2 = test()
+    assert not equal
+    assert inside
+    assert inside2
+    assert not outside
+    assert inside21
+    assert inside22
+    assert not outside2
+
+
+@pytest.mark.parametrize('graph', [True, False])
+@pytest.mark.parametrize('testclass', [Limit, lambda limit_fn, rect_limits: zfit.Space('obs1', limits=limit_fn,
+                                                                                       rect_limits=rect_limits)])
+def test_limits_1d(graph, testclass):
+    def test(allow_graph=True):
+        limit = testclass(limit_fn=inside_rect_limits, rect_limits=rect_limits)
+        limit2 = testclass(limit_fn=inside_rect_limits, rect_limits=rect_limit_enlarged)
+
+        assert not limit.has_rect_limits
+        assert limit.has_limits
+        assert not limit.limits_not_set
+        assert not limit.limits_are_false
+        inside = limit.inside(2)
+        inside2 = limit.inside(2, guarantee_limits=True)
+        outside = limit.inside(4)
+
+        assert not limit2.has_rect_limits
+        assert limit2.has_limits
+        assert not limit2.limits_not_set
+        assert not limit2.limits_are_false
+        inside21 = limit2.inside(2)
+        inside22 = limit2.inside(2, guarantee_limits=True)
+        outside2 = limit2.inside(4)
+        if graph:
+            equal = limit.equal(limit2, allow_graph=allow_graph)
+        else:
+            equal = limit == limit2
+
+        return inside, inside2, outside, equal, inside21, inside22, outside2
+
+    if graph:
+        test = z.function(test)
+    inside, inside2, outside, equal, inside21, inside22, outside2 = test()
+    assert not equal
+    assert inside
+    assert inside2
+    assert not outside
+    assert inside21
+    assert inside22
+    assert not outside2
```

Comment:
I'll make this comment only for this test case but it applies to all test cases in this file.
Why do you split this into a function `test()` and a block of assertions (line 133-139)?

If one of the assertions fails, I have to search first search for the related code to find it.

Just write the assertion on the line where you would assign the variable. Most of your expression are self-explanatory and don't need to be named by a variable. Example:
Do
```python
assert limit.inside(2)
```
instead of 
```python
inside = limit.inside(2)
# ... twenty lines of code
assert inside
```
This way you can replace the function `test()` by it's body, i.e. delete about 15 lines of code and make it more concise. Do not try to generalize stuff in advance! One of the best guidelines I learnt in my internship is the rule of three. Only refactor stuff into a function (or other general procedure) once you encounter the same code three 
times. Otherwise your make premature assumptions of what needs to be generalized, which ultimately leads to more refactoring and more code. 
https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)

Reply:
> One of the best guidelines I learnt in my internship is the rule of three.

Yes, I agree, just recently learned that too with backends. Either you support 1, 2 or many. One is specific, for two you may still hack. But if you support 3, it's general enough.

> Why do you split this into a function test() and a block of assertions (line 133-139)?

This is because the test is once wrapped within a `tf.function` and once not. This makes it run once _inside a graph context_ and execute (e.g. the 'equal' that is returned is calculated in graph mode).

Some things should hold inside the graph mode, some should be returned correctly (because then they are evaluated).

So the setup, while I agree not great (there is no easy way to tell pytest that it should go also one level deeper to check failed tests? E.g. if an assert inside `test()` fails, it would be good if it points inside the function).